proxy mode for http proxies (#734)
encode/httpcore: docs/proxies.md

)
```

## Proxy Connection Mode

There are two types of HTTP proxy:

> 1. Forwading (HTTP [absolute-url](https://tools.ietf.org/html/rfc7230#section-5.3.2))

Author: ```suggestion
1. Forwarding (HTTP [absolute-url](https://tools.ietf.org/html/rfc7230#section-5.3.2))
```

2. Tunneling (HTTP [`CONNECT` method](https://tools.ietf.org/html/rfc7231#section-4.3.6))

> By default we forwarding plain http requests and tunneling https requests.

Reviewer: ```suggestion
By default `httpcore` will use forwarding for http requests and tunneling for https requests.
```

> You can change this behavior with `proxy_mode: httpcore.ProxyMode` parameter:

Reviewer: ```suggestion

You can change this behavior with `proxy_mode: httpcore.ProxyMode` parameter:

```

```py
import httpcore
import base64

proxy = httpcore.HTTPProxy(
    proxy_url="http://127.0.0.1:8080/",

>     proxy_mode=httpcore.ProxyMode.TUNNEL,  # Forces HTTP requests also use Tunneling

Reviewer: I guess let's just drop the comment here, since the docs are already giving us context...

```suggestion
    proxy_mode=httpcore.ProxyMode.TUNNEL,
```

)
```

## Proxy Connection Mode

There are two types of HTTP proxy:

1. Forwading (HTTP [absolute-url](https://tools.ietf.org/html/rfc7230#section-5.3.2))
2. Tunneling (HTTP [`CONNECT` method](https://tools.ietf.org/html/rfc7231#section-4.3.6))

By default we forwarding plain http requests and tunneling https requests.
You can change this behavior with `proxy_mode: httpcore.ProxyMode` parameter:
```py
import httpcore
import base64

proxy = httpcore.HTTPProxy(
    proxy_url="http://127.0.0.1:8080/",
    proxy_mode=httpcore.ProxyMode.TUNNEL,  # Forces HTTP requests also use Tunneling
)
```

Note that `ProxyMode.FORWARD` will enable forwarding https requests, so they will be visible for proxy server.
It means handling TLS stuffs like certificate validation would on proxy side.

> 

Reviewer 2: ```suggestion

```

)
```

## Proxy Connection mode

There are two types of HTTP proxy:

1. Forwading (using absolute-url)
2. Tunneling (using `CONNECT` method)

By default we forwarding plain http requests and tunneling https requests.
You can change this behavior with `mode: httpcore.ProxyMode` parameter:
```py
import httpcore
import base64

proxy = httpcore.HTTPProxy(
    proxy_url="http://127.0.0.1:8080/",
    mode=httpcore.ProxyMode.HTTP_TUNNEL,  # Forces HTTP requests also use Tunneling
)
```
`ProxyMode` is a flag parameter, also you can reverse current default behavior:
```py
import httpcore
import base64

proxy = httpcore.HTTPProxy(
    proxy_url="http://127.0.0.1:8080/",

>     mode=httpcore.ProxyMode.HTTP_TUNNEL | httpcore.ProxyMode.HTTPS_FORWARD,

Author: by default, plain http requests use forwarding method to connect proxy server, and https requests use tunneling.
this will reverse the default.

>     mode=httpcore.ProxyMode.HTTP_TUNNEL | httpcore.ProxyMode.HTTPS_FORWARD,

Reviewer: > this will reverse the default

I can't see any reason we'd want to support this.

>     mode=httpcore.ProxyMode.HTTP_TUNNEL | httpcore.ProxyMode.HTTPS_FORWARD,

Author: I tried to cover all possibilities,
So, as it is useless, should I remove it?

>     mode=httpcore.ProxyMode.HTTP_TUNNEL | httpcore.ProxyMode.HTTPS_FORWARD,

Reviewer: Yes let's not support that as a use-case.

>     mode=httpcore.ProxyMode.HTTP_TUNNEL | httpcore.ProxyMode.HTTPS_FORWARD,

Author: done in https://github.com/encode/httpcore/pull/734/commits/067d76f3c92611629d768d01cb5f03eddb53b894

)
```

> (WARN: `ProxyMode.HTTPS_FORWARD` is unsafe, use it when you trust proxy server)

Reviewer: how do other texts phrase this? can we explain the implications to readers more clearly?

Author: perhaps here needs fully explain that what is forwarding and tunneling proxy.
